src/main/java/mekanism/common/content/gear/mekatool/ModuleVeinMiningUnit.java

```diff
@@ -49,30 +51,36 @@ public int getExcavationRange() {
         return excavationRange.get().getRange();
     }
 
-    public static Set<BlockPos> findPositions(BlockState state, BlockPos location, Level world, int maxRange) {
-        Set<BlockPos> found = new LinkedHashSet<>();
-        Set<BlockPos> openSet = new LinkedHashSet<>();
-        openSet.add(location);
-        Block startBlock = state.getBlock();
-        int maxCount = MekanismConfig.gear.disassemblerMiningCount.get() - 1;
-        while (!openSet.isEmpty()) {
-            BlockPos blockPos = openSet.iterator().next();
-            found.add(blockPos);
+    public static Map<BlockPos, BlockState> findPositions(Level world, Map<BlockPos, BlockState> initial, boolean extended, int maxRange) {
+        int maxVein = MekanismConfig.gear.disassemblerMiningCount.get();
+        int maxCount = (int) (initial.size() + maxVein * initial.values().stream().map(bs->bs.getBlock()).distinct().count());
+        Map<BlockPos, BlockState> found = new LinkedHashMap<>();
+        Map<BlockPos, BlockState> openSet = new LinkedHashMap<>();
+        openSet.putAll(initial);
+        for (int dist = 0, next = openSet.size(); !openSet.isEmpty(); next += (found.size() == next && ++dist > 0) ? openSet.size() : 0) {
+            Entry<BlockPos, BlockState> blockEntry = openSet.entrySet().iterator().next();
+            BlockPos blockPos = blockEntry.getKey();
+            BlockState blockState = blockEntry.getValue();
+            found.put(blockPos, blockState);
             openSet.remove(blockPos);
             if (found.size() > maxCount) {
-                return found;
+                break;
             }
-            for (BlockPos pos : BlockPos.betweenClosed(blockPos.offset(-1, -1, -1), blockPos.offset(1, 1, 1))) {
-                //We can check contains as mutable
-                if (!found.contains(pos) && (maxRange == -1 || WorldUtils.distanceBetween(location, pos) <= maxRange)) {
-                    Optional<BlockState> blockState = WorldUtils.getBlockState(world, pos);
-                    if (blockState.isPresent() && startBlock == blockState.get().getBlock()) {
-                        //Make sure to add it as immutable
-                        if (openSet.add(pos.immutable())) {
+            boolean isOre = blockState.is(MekanismTags.Blocks.ATOMIC_DISASSEMBLER_ORE);
```

Comment:
Given we may be operating on a lot of blocks and in large modded environments blocks may have lots and lots of tags, I wonder if this would be more performant as having a map outside of the for loop and then implement calculating isOre like this:
```java
//types is the calculated distinct amount portion of maxCount
Object2BooleanMap<Block> oreTracker = new Object2BooleanArrayMap<>(types);
...
boolean isOre = oreTracker.computeIfAbsent(block, (Block b) -> b.builtInRegistryHolder().is(MekanismTags.Blocks.ATOMIC_DISASSEMBLER_ORE));
```

And then potentially doing the same for calculating power usage in the meka tool?